### DISCUSSION TITLE
How to use ExtensionSettings option

### DISCUSSION THREAD

**Discussion by Mithrandir918** [mozilla/policy-templates#835] [Q&A]
Hi! I'm sorry for the trouble, but I do not quite understand how to use this. Specifically, I'm trying to use the ExtensionSettings command to prevent me from uninstalling some extensions. Could someone please try and walk me through this? Thank you!

**Comment by mkaply** [comment]
What extensions are you wanting to do this with?

**Comment by Mithrandir918** [comment]
I’ll link it. Thanks for helping me out! https://addons.mozilla.org/en-US/firefox/addon/blocksite/

Sent with a Spark

**Comment by mkaply** [comment]
I'll take you through the steps I do. So the first thing we need is the ID of the extension. There are a few ways to do that.

1. Install the extension and go to about:debugging
2. Unzip the extension and look at manifest.json (doesn't always work).
3. Install the extension I specifically built to do this - https://github.com/mkaply/queryamoid/releases

Using my tool I got the ID:

{dd3d7613-0246-469d-bc65-2a3cc1668adc}

and the download URL:

https://addons.mozilla.org/firefox/downloads/latest/blocksite/latest.xpi

So what you're looking for is "force_installed" that says "install this for the user and don't let them uninstall.

So you would set ExtensionSettings to:

```
{
  "{dd3d7613-0246-469d-bc65-2a3cc1668adc}": {
    "installation_mode": "force_installed",
    "install_url": "https://addons.mozilla.org/firefox/downloads/latest/blocksite/latest.xpi"
  }
}
```

**Reply by Mithrandir918** [reply]
Awesome. Where in the firefox installation do I put this file?

**Reply by Mithrandir918** [reply]
I got it into the distribution folder (which is in FireFox's package contents), but it's still not doing anything. The policies.json file is in the distribution folder. However, when I click on it, firefox says "SyntaxError: JSON.parse: expected property name or '}' at line 1 column 2 of the JSON data". Not being a coder, I'm not sure what that means. Do I need to make a new installation of firefox? If it helps at all, I'm on Mac. If you could help me, I'd really appreciate it. Thank you!

**Comment by mkaply** [comment]
I would recommend linting your file at jsonlint.com. It should be a matter of putting this file called policies.json in the distribution directory:

```
{
  "policies": {
    "ExtensionSettings": {
      "{dd3d7613-0246-469d-bc65-2a3cc1668adc}": {
        "installation_mode": "force_installed",
        "install_url": "https://addons.mozilla.org/firefox/downloads/latest/blocksite/latest.xpi"
      }
    }
  }
}
```

**Reply by Mithrandir918** [reply]
I did all of that. When I click on the file in the distribution folder, it does not come up with an error message like it did before. However, I can still remove the extension. I really appreciate the help.

**Comment by mkaply** [comment]
IS the policy active when you go to about:policies?

**Reply by Mithrandir918** [reply]
It is not.

**Reply by mkaply** [reply]
OK, so it sounds like it's not reading the file. Can you post the policies.json? What platform are you on?

**Comment by mkaply** [comment]
Oh  mac :). OK the location is different.

Are you putting it in 

 Firefox.app/Contents/Resources/distribution

**Comment by Mithrandir918** [comment]
Yes. I created the distribution folder though because it wasn’t there before. Is that the problem?

Sent with a Spark

**Comment by mkaply** [comment]
> Yes. I created the distribution folder though because it wasn’t there before. Is that the problem?

Nope, that should work. I'm at a loss. Can you post a screenshot of your about:policies?

**Reply by Mithrandir918** [reply]
<img width="1440" alt="Screen Shot 2021-12-07 at 11 37 24 AM" src="https://user-images.githubusercontent.com/95041308/145069411-4fdb2653-8891-44bc-9fdf-a4e7cfc69ee6.png">
Is this what you're looking for?

**Comment by mkaply** [comment]
Yes.

so when you click on Errors on the left, what does it say?

**Reply by Mithrandir918** [reply]
Error parsing JSON file: SyntaxError: JSON.parse: expected property name or '}' at line 1 column 2 of the JSON data

I think that might be an old log though from when I was trying to do it myself.

**Comment by mkaply** [comment]
So there's definitely still an error in the policies.json. Can you post it here?

**Comment by Mithrandir918** [comment]
Error parsing JSON file: SyntaxError: JSON.parse: expected property name or '}' at line 1 column 2 of the JSON data

I think that have been from me messing with it beforehand before I asked you to help.

Sent with a Spark

**Comment by Mithrandir918** [comment]
Yep. I can’t post a .json, so I’ll send a screenshot

Sent with a Spark

**Comment by mkaply** [comment]
Screenshot didn't make it. If you want to email it to me - mkaply at mozilla dot com

**Reply by Mithrandir918** [reply]
I'll try again here. 
<img width="774" alt="Screen Shot 2021-12-07 at 11 49 11 AM" src="https://user-images.githubusercontent.com/95041308/145071807-9d954be1-455a-45de-84e7-83b8786aef0c.png">

**Comment by mkaply** [comment]
I feel like it's reading a different policies.json file than that one because that one is definitely correct.

**Reply by Mithrandir918** [reply]
Huh. How would I check for that? There are no other files in the distributions folder.

**Reply by Mithrandir918** [reply]
Should I delete and reinstall Firefox with the folder in there?

**Comment by mkaply** [comment]
Can you try just deleting policies.json and see what you get in about:policies?

I just want to make sure it's that file being used.

**Reply by Mithrandir918** [reply]
Sorry, had to go. It says "The Enterprise Policies service is inactive."

**Reply by mkaply** [reply]
Is there still an error tab?

**Reply by Mithrandir918** [reply]
Yes, the same message is still there

**Comment by mkaply** [comment]
OK, that tells me it's reading a different policies.json. Can you go to

about:support and tell me the value for Application Binary?

**Reply by Mithrandir918** [reply]
I'm not exactly sure what you mean, but it says this:  	/Applications/Firefox.app/Contents/MacOS/firefox

**Reply by Mithrandir918** [reply]
I see that file. Should I delete it?

**Reply by mkaply** [reply]
Can you check if there is a policies.json in 

 /Applications/Firefox.app/Contents/Resources/distribution?

**Reply by Mithrandir918** [reply]
There is not.

**Comment by mkaply** [comment]
I'm at a loss here. Would you be willing to have a video call/screen share where we try to diagnose?

**Comment by Mithrandir918** [comment]
Thank you for offering, but I don’t feel comfortable doing that. Thank you for your help though! I really appreciate the time you’ve put in helping me.

Sent with a Spark

**Comment by mkaply** [comment]
Totally understand. My only other suggestion is to go to about:config and add a preference called

browser.policies.loglevel

and set the value to a string with "debug"

Then the JS console ( Cmd + Shift + J) might give us some extra info.

The other thing I was wondering is if you somehow have a configuration profile loaded (separate from policies.json)

**Reply by Mithrandir918** [reply]
How do I set the string value to "debug"? I'm not seeing the option there.

**Comment by Mithrandir918** [comment]
I’ll try to add the preference. Where would I look for a different config file outside of policies.json?

Sent with a Spark

**Comment by mkaply** [comment]
Go to about:config

Type browser.policies.loglevel in the field.

Click the string radio button

Click the Plus sign

type the word debug and press enter.

As far as having a configuration profile applied, search your mac using the question mark in the upper right for "Profiles"  You should get a hit that is in system preferences. It's a checkbox in a star. Open that up and see if there is anything in it (it will be blank if you have nothing applied)

**Reply by Mithrandir918** [reply]
There is nothing in profiles. I set up the browser.policies.loglevel thing. What should I look for in the debug screen?

**Comment by mkaply** [comment]
When you first open Firefox, you'll see the same JSON error in the console, but you should see some extra text before that (maybe). Basically a screenshot of the JS console right after starting Firefox.

Unfortunately this won't have everything we need, but I'm going to add some better debug info just in case this ever happens again.

**Reply by Mithrandir918** [reply]
Got it. So, just to clarify, I should restart Firefox and send a screenshot of the debug screen right after booting it?

**Reply by mkaply** [reply]
Exactly.

**Reply by Mithrandir918** [reply]
<img width="907" alt="Screen Shot 2021-12-08 at 12 48 59 PM" src="https://user-images.githubusercontent.com/95041308/145258169-3505c59c-f1fc-4735-b82c-8b3be974b942.png">

**Comment by mkaply** [comment]
Yeah, nothing that would help. I'm putting a change in Firefox 96 that will help us debug this.

**Reply by Mithrandir918** [reply]
Thanks!

**Reply by Mithrandir918** [reply]
Do you think the issue will be fixed when Firefox 96 is released?

**Comment by mkaply** [comment]
I don't think the issue will be fixed, but I'll make sure there's better info so we can debug the problem.

96 is in a month.

**Comment by Mithrandir918** [comment]
Ok. Thanks for helping me!

Sent with a Spark

**Comment by mkaply** [comment]
So just to followup what I'm doing here. I'm putting a patch into firefox 96 that shows the path and content of policies.json.

So when Firefox 96 is released, we can revisit and figure out where the file is being read on your machine. So I'll reping you beginning of january.

**Comment by Mithrandir918** [comment]
Awesome! Send me a message when that’s up please!

Sent with a Spark

**Comment by angela-d** [comment]
@Mithrandir918 if you manage your Macs with MDM, like JAMF or Filewave, you can simply deploy a mobileconfig policy to them.
I have a copy of what I use in my notes, maybe it'll help: https://github.com/angela-d/brain-dump/tree/master/sysadmin/macos/mdm-profiles

**Comment by Mithrandir918** [comment]
And I got that working! Thanks!

Sent with a Spark
